### ISSUE TITLE
Input balance per delay line

### ISSUE THREAD

**Issue by Bleuzen** [lsp-plugins/lsp-plugins#404]
Hi, in the stereo delay plugin I want to set up delay lines sometimes only based on one channel
so for example delay line 0 should be based on the left channel
and delay line 1 should be based on the right channel

currently I do not see how this is possible. There are panorama settings but even if both are set to the left like this
![](https://github.com/lsp-plugins/lsp-plugins/assets/12885163/541b3464-de19-4ed4-9337-c480f24398df)
The delay will still contain the right input channel (only routed to the left)
So in this example with the existing settings there seems no way to eliminate the right channel for this delay line

My first idea was to add gain knobs per input channel. I already experimented with it and this is how it looks:
![](https://github.com/lsp-plugins/lsp-plugins/assets/12885163/73eef895-e419-470d-9015-c64535c7e455)

This seems pretty cool but does add two knobs in each row so I'm not sure if the increase in window size is worth the benefit

However thinking about it this could also work with one knob. Maybe something like this (naming not finalized, open for ideas)
![](https://github.com/lsp-plugins/lsp-plugins/assets/12885163/6382c29f-9f26-477a-bde4-ceb8bcd86959)

Logic:
if it is set to `-100` (like in the screenshot) the delay line will only use the left input channel
if set to `100` the delay line will only use the right input channel
if set to `0` it will feed both input channels equally into the delay line (like the plugin behaves currently)

What do you think about this solution (and naming)?

**Comment by Bleuzen** [comment]
Sorry about the mess with the edits, it is late and I'm still learning audio stuff...
just figured out what I was talking about is called stereo balance I guess

**Comment by sadko4u** [comment]
Hi!
Yes, we already have the `Gain` knob at the output of the delay line, so the `Balance` knob seems to be reasonable solution.

**Comment by sadko4u** [comment]
I think we can introduce the balance knob similar to how it is introduced in `Parametric equalizer` plugin: If the knob is in the leftmost position, then the left channel is doubled (+6 dB) and right channel is zeroed (0 dB). Such behaviour allows us to introduce the simpliest possible formula for the left and right channels:
```
left_gain = (100 - balance) * 0.01  // 2 at -100, 1 at 0, 0 at 100
right_gain = (100 + balance) * 0.01 // 0 at -100, 1 at 0, 2 at 100
```

**Comment by Bleuzen** [comment]
> so the `Balance` knob seems to be reasonable solution

yeah sounds good


> I think we can introduce the balance knob similar to how it is introduced in Parametric equalizer plugin: If the knob is in the leftmost position, then the left channel is doubled (+6 dB) and right channel is zeroed (0 dB).

Oh I didn't know it is already implemented like this in Parametric equalizer. Tbh I'm not a fan of it.
Especially of this:

> channel is doubled (+6 dB)

because this is annoying to deal with. You always have to re-do the levelling in the mix / adjust the gain after adjusting the balance and it can introduce clipping / unwanted distortion

It is even worse thinking about automation like this:
- start with the balance know at 0 (both channels +-0db)
- over time, move the balance knob to the left
- now the signal becomes louder so we have to add another automation to the gain knob to account for this
- over time, move the balance knob from the left to the right
- we have to do another gain automation from -6 to 0 (while the balance is moving from -100 to 0) and then again from 0 to -6 (while the balance is moving from 0 to 100)

Idk maybe I'm missing something but at first thought doubling a channel seems not exactly user friendly

I looked at other software and it does never increase the volume of a channel, only attenuate the other channel

Under the music players, DAWs and other plugins tested so far there seems to be some consensus about how balance should work - LSP Plugins is the outlier here

For example this is from the balance plugin from x42 (one of the Ardour devs):
![](https://github.com/lsp-plugins/lsp-plugins/assets/12885163/3d018605-eb1a-4b2f-b2e8-3eb3aefe2251)

When the balance is set 50% to the right, the right channel will stay as is and the left channel is at -6.0dB

From a users perspective this seems the most friendly to me - at first thought of my audio noob brain - feel free to correct me if I'm wrong




> Such behaviour allows us to introduce the simpliest possible formula for the left and right channels

Just curious, does this have any noticeable performance (or other) benefit or did you mean it is just the simpliest to code from a dev perspective?

**Comment by Bleuzen** [comment]
> When the balance is set 50% to the right, the right channel will stay as is and the left channel is at -6.0dB

While this still sounds the best to work with to me, maybe...

- note that it is inconsistent with the way balance is implemented in the LSP parametric equalizer
- check how this would work in combination with the existing panorama controls

(mostly just writing down notes for now as I have not fully formed an opinion on this yet)

**Comment by sadko4u** [comment]
OK, then the formula can be changed to:
```
left_gain = min((100 - balance) * 0.01, 1)  // 1 at -100, 1 at 0, 0 at 100
right_gain = min((100 + balance) * 0.01, 1) // 0 at -100, 1 at 0, 1 at 100
```

**Comment by sadko4u** [comment]
Just pushed all necessary changes to `devel` branch.

**Comment by sadko4u** [comment]
Available in 1.2.17.
